ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/server/resources/streaming/PullQueryStreamWriterTest.java

```diff
@@ -0,0 +1,268 @@
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+import org.mockito.stubbing.Answer;
+
+@SuppressWarnings("unchecked")
+@RunWith(MockitoJUnitRunner.class)
+public class PullQueryStreamWriterTest {
+
+  private static final LogicalSchema SCHEMA = LogicalSchema.builder()
+      .valueColumn(ColumnName.of("a"), SqlTypes.STRING)
+      .build();
+
+  @Rule
+  public final Timeout timeout = Timeout.builder()
+      .withTimeout(30, TimeUnit.SECONDS)
+      .withLookingForStuckThread(true)
+      .build();
+
+  @Mock
+  private PullQueryResult pullQueryResult;
+  @Mock
+  private PullQueryQueue pullQueryQueue;
+  @Mock
+  private Clock clock;
+
+  @Captor
+  private ArgumentCaptor<Consumer<Throwable>> throwableConsumerCapture;
+  @Captor
+  private ArgumentCaptor<Consumer<Void>> limitCapture;
+
+  private ByteArrayOutputStream out;
+
+
+  private PullQueryStreamWriter writer;
+
+  @Before
+  public void setUp() {
+    when(pullQueryResult.getQueryId()).thenReturn(new QueryId("Query id"));
+    when(pullQueryResult.getSchema()).thenReturn(SCHEMA);
+    doNothing().when(pullQueryResult).onCompletion(limitCapture.capture());
+    when(pullQueryQueue.isEmpty()).thenReturn(false);
+    writer = new PullQueryStreamWriter(pullQueryResult, 1000, ApiJsonMapper.INSTANCE.get(),
+        pullQueryQueue, clock, new CompletableFuture<>());
+
+    out = new ByteArrayOutputStream();
+  }
+
+  @Test
+  public void shouldWriteAnyPendingRowsBeforeReportingException() {
+    // Given:
+    doAnswer(streamRows("Row1", "Row2", "Row3"))
+        .when(pullQueryQueue).drainRowsTo(any());
+
+    givenUncaughtException(new KsqlException("Server went Boom"));
+
+    // When:
+    writer.write(out);
+
+    // Then:
+    final List<String> lines = getOutput(out);
+    assertThat(lines, contains(
+        containsString("header"),
+        containsString("Row1"),
+        containsString("Row2"),
+        containsString("Row3"),
+        containsString("Server went Boom")
+    ));
+  }
+
+  @Test
+  public void shouldWriteNoRows() {
+    // Given:
+    when(pullQueryQueue.isClosed()).thenReturn(true);
+    doAnswer(streamRows())
+        .when(pullQueryQueue).drainRowsTo(any());
+
+    // When:
+    writer.write(out);
+
+    // Then:
+    final List<String> lines = getOutput(out);
+    assertThat(lines, contains(
+        containsString("header")
+    ));
+  }
+
+  @Test
+  public void shouldExitAndDrainIfPullQueryQueueIsClosed() {
+    // Given:
+    when(pullQueryQueue.isClosed()).thenReturn(true);
+    doAnswer(streamRows("Row1", "Row2", "Row3"))
+        .when(pullQueryQueue).drainRowsTo(any());
+
+    // When:
+    writer.write(out);
+
+    // Then:
+    final List<String> lines = getOutput(out);
+    assertThat(lines, contains(
+        containsString("header"),
+        containsString("Row1"),
+        containsString("Row2"),
+        containsString("Row3")));
+  }
+
+  @Test
+  public void shouldExitAndDrainIfLimitReached() {
```

Comment:
should we have negative tests for this and the above? i.e. does not drain if limit not reached or pull query not closed?